File: 22_generate_parentheses/generateParenthesis.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        result = []
        def generateStr(nleft, nright, s):
            if nleft == n and nright == n:
                result.append(s)

            if nleft <= n:
                generateStr(nleft + 1, nright, s + '(')

            if nright < nleft:
                generateStr(nleft, nright + 1, s + ')')



        generateStr(0, 0, '')
        return result

sol = Solution()
n = 15
print sol.generateParenthesis(n)

File: 22_generate_parentheses/generateParenthesis.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        dfs(n, "", 0, 0, res);
        return res;
    }
private:
    void dfs(int n, string s, int nleft, int nright, vector<string>& res) {
        if(nleft == n && nright == n) {
            res.push_back(s);
            return;
        }

        if(nleft < n)
            dfs(n, s + '(', nleft + 1, nright, res);
        if(nright < nleft) // If nright > nleft, then it must be invalid
            dfs(n, s + ')', nleft, nright + 1, res);
    }
};
